### ISSUE TITLE
Get Literals and Proof Eq vars for ProofOnlySimpleIntegerVariableID

### ISSUE THREAD

**Issue by mmcilree** [ciaranm/glasgow-constraint-solver#27]
What's the best way to do this?
- Expand what is allowed for LiteralFromIntegerVariableID?
- Make something like LiteralFromProofOnlyIntegerVariableID?
- Bypass Literals entirely and just define some kind of need_direct_eq_var(ProofOnlyIntegerVariableID var, Integer val)

**Comment by ciaranm** [comment]
Where do you need to use the literals? I think the answer might be to expand anywhere that can reasonably take either a proper variable or a proof variable to have a broader API.

**Comment by mmcilree** [comment]
Okay, in that case perhaps changing LiteralFromIntegerVariableID should be okay. 

In `set_up_direct_only_variable_encoding` it looks like it might be useful.

https://github.com/ciaranm/glasgow-constraint-solver/blob/a2b4982b67350e3824b185404b6d3066c88598d8/gcs/innards/proof.cc#L280-L288

Also perhaps we generalise `need_direct_encoding_for` to allow `ProofOnlyIntegerVariable`s, but this does the check: 

https://github.com/ciaranm/glasgow-constraint-solver/blob/a2b4982b67350e3824b185404b6d3066c88598d8/gcs/innards/proof.cc#L482-L483

So again we probably need to be able to make literals out of them

**Comment by ciaranm** [comment]
I don't want the full Literal API to allow for ProofOnly things, because this will leak out of innards and into things in the user-facing API. I'm trying to avoid cases where you can ask for something that won't work, because (as you noticed...) it gets a bit odd if something compiles and then doesn't do the right thing. The way I'd tackle this is to work out which exactly functions you want to call with ProofOnly things, and then from there to carefully extend with variants. For example, for `direct_integer_variables` this probably doesn't want to be a map from LiteralFromIntegerVariableID any more, but rather from something like a LiteralFromIntegerOrProofOnlyVariableID.

The nice thing about variant, compared to traditional inheritance, is it lets us do this all over the place fairly easily and to be very specific in what's supported where.

**Comment by mmcilree** [comment]
Yes this looks useful. Shall I mess around on the same branch to try to allow creation of literals for ProofOnlySimpleIntegerVariableID -- that's the main thing that's needed to progress with CircuitSCC -- since I want to log a WeightPBSum like pos_a=b + pos_c=d + pos_e=f >= 1;. Or shall I leave you to do that?

**Comment by mmcilree** [comment]
Okay sure. Bear in mind nasty workaround for clang the https://github.com/ciaranm/glasgow-constraint-solver/blob/f6bede0c5cac30b84bf2a0ddf4702d5f667601ac/gcs/variable_id.hh#L91C1-L91C1

I immediately ran into this when hacking around (on a discard branch) for a new literal type earlier.

**Comment by ciaranm** [comment]
The rup-api branch now passes all tests, but could well still have bugs, and it probably misses some features that you need, but you should be able to at least get things to compile if you put a ProofVariable inside a PB sum. I guess you also want to put literals on ProofVariables inside there too?

**Comment by ciaranm** [comment]
I'm now wondering why Literal is even in the public API, rather than the innards.

**Comment by mmcilree** [comment]
Great. Just logistically speaking, probably the best way to evaluate this is to somehow get the changes into my circuit-scc branch and try it out with what I'm trying to do. In terms of keeping clean commit history and making it nice when we eventually merge this all into main what's the best way to go about this?

Merge rup-api into circuit-scc? 
Rebase? 
Merge rup-api into main first? 
Make another branch circuit-scc-and-rup-api?

**Comment by ciaranm** [comment]
If it doesn't look too outrageously wrong to you, I could just merge the
branch to main.

**Comment by ciaranm** [comment]
It is now merged, and probably won't throw UnimplementedException any more, but I've not tried it or tested it so it probably doesn't actually work.

**Comment by mmcilree** [comment]
Okay I have hit an unimplemented exception. Will discuss in meeting - gonna get some coffee first.
